homeassistant/components/recorder/migration.py

```diff
@@ -821,10 +821,88 @@ def _apply_update(  # noqa: C901
         # Once we require SQLite >= 3.35.5, we should drop the column:
         # ALTER TABLE statistics_meta DROP COLUMN state_unit_of_measurement
         pass
+    elif new_version == 31:
+        # Once we require SQLite >= 3.35.5, we should drop the column:
+        # ALTER TABLE events DROP COLUMN time_fired
+        # ALTER TABLE states DROP COLUMN last_updated
+        # ALTER TABLE states DROP COLUMN last_changed
+        _add_columns(session_maker, "events", ["time_fired_ts FLOAT"])
+        _add_columns(
+            session_maker, "states", ["last_updated_ts FLOAT", "last_changed_ts FLOAT"]
+        )
+        _create_index(session_maker, "events", "ix_events_time_fired_ts")
+        _create_index(session_maker, "events", "ix_events_event_type_time_fired_ts")
+        _create_index(session_maker, "states", "ix_states_entity_id_last_updated_ts")
+        _create_index(session_maker, "states", "ix_states_last_updated_ts")
+        with session_scope(session=session_maker()) as session:
+            _migrate_columns_to_timestamp(hass, session, engine)
+    elif new_version == 32:
+        # Migration is done in two steps to ensure we can start using
+        # the new columns before we wipe the old ones.
+        with session_scope(session=session_maker()) as session:
+            _wipe_old_string_time_columns(hass, session, engine)
+        _drop_index(session_maker, "states", "ix_states_entity_id_last_updated")
+        _drop_index(session_maker, "events", "ix_events_event_type_time_fired")
+        _drop_index(session_maker, "states", "ix_states_last_updated")
+        _drop_index(session_maker, "events", "ix_events_time_fired")
     else:
         raise ValueError(f"No schema migration defined for version {new_version}")
 
 
+def _wipe_old_string_time_columns(
+    hass: HomeAssistant, session: Session, engine: Engine
+) -> None:
+    """Wipe old string time columns."""
+    # Wipe Events.time_fired
+    # Wipe States.last_updated
+    # Wipe States.last_changed
```

Comment:
```suggestion
    # Wipe Events.time_fired since its been replaced by Events.time_fired_ts
    # Wipe States.last_updated since its been replaced by States.last_updated_ts
    # Wipe States.last_changed since its been replaced by States.last_changed_ts
```